iot_packet_verifier/src/verifier.rs

```diff
@@ -0,0 +1,490 @@
+        Ok(*balance >= amount)
+    }
+}
+
+#[async_trait]
+pub trait ConfigServer {
+    type Error;
+
+    async fn fetch_org(
+        &mut self,
+        oui: u64,
+        cache: &mut HashMap<u64, PublicKeyBinary>,
+    ) -> Result<PublicKeyBinary, Self::Error>;
+
+    async fn enable_org(&mut self, oui: u64) -> Result<(), Self::Error>;
+
+    async fn disable_org(&mut self, oui: u64) -> Result<(), Self::Error>;
+}
+
+// TODO: Move this somewhere else
+
+// Probably should change name to something like OrgClientCache to be more
+// consistent with BalanceCache
+pub struct CachedOrgClient {
+    pub keypair: Keypair,
+    pub enabled_clients: HashMap<u64, bool>,
+    pub client: OrgClient<Channel>,
+}
+
+impl CachedOrgClient {
+    pub fn new(client: OrgClient<Channel>, keypair: Keypair) -> Self {
+        CachedOrgClient {
+            keypair,
+            enabled_clients: HashMap::new(),
+            client,
+        }
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum OrgClientError {
+    #[error("Rpc error: {0}")]
+    RpcError(#[from] tonic::Status),
+    #[error("Crypto error: {0}")]
+    CryptoError(#[from] helium_crypto::Error),
+}
+
+#[async_trait]
+impl ConfigServer for CachedOrgClient {
+    type Error = OrgClientError;
+
+    async fn fetch_org(
+        &mut self,
+        oui: u64,
+        cache: &mut HashMap<u64, PublicKeyBinary>,
+    ) -> Result<PublicKeyBinary, Self::Error> {
+        if !cache.contains_key(&oui) {
+            let req = OrgGetReqV1 { oui };
+            let pubkey =
+                PublicKeyBinary::from(self.client.get(req).await?.into_inner().org.unwrap().owner);
+            cache.insert(oui, pubkey);
+        }
+        Ok(cache.get(&oui).unwrap().clone())
+    }
+
+    async fn enable_org(&mut self, oui: u64) -> Result<(), Self::Error> {
+        if !mem::replace(self.enabled_clients.entry(oui).or_insert(false), true) {
+            let mut req = OrgEnableReqV1 {
+                oui,
+                timestamp: Utc::now().timestamp_millis() as u64,
+                signer: self.keypair.public_key().to_vec(),
+                signature: vec![],
+            };
+            let signature = self.keypair.sign(&req.encode_to_vec())?;
+            req.signature = signature;
+            let _ = self.client.enable(req).await?;
+        }
+        Ok(())
+    }
+
+    async fn disable_org(&mut self, oui: u64) -> Result<(), Self::Error> {
+        if mem::replace(self.enabled_clients.entry(oui).or_insert(true), false) {
+            let mut req = OrgDisableReqV1 {
+                oui,
+                timestamp: Utc::now().timestamp_millis() as u64,
+                signer: self.keypair.public_key().to_vec(),
+                signature: vec![],
+            };
+            let signature = self.keypair.sign(&req.encode_to_vec())?;
+            req.signature = signature;
+            let _ = self.client.disable(req).await?;
+        }
+        Ok(())
+    }
+}
+
+#[async_trait]
+pub trait Burner {
```

Comment:
are some of these traits not in use yet? the Burner trait at least seems to be ignored by the Burner struct when it runs its loop

Reply:
This is in use, it's different from the `Burner` struct. This trait is for inserting the burn amount into `pending_burns`, I just wasn't sure what to name it other than giving it the same name. Perhaps I'll rename the struct to `BurnDaemon` or something. This trait is used so that I can mock the database in unit tests